src/traversals/dfs.jl

```diff
@@ -87,32 +86,31 @@ function topological_sort_by_dfs end
 @traitfn function topological_sort_by_dfs(g::AG::IsDirected) where {T,AG<:AbstractGraph{T}}
     vcolor = zeros(UInt8, nv(g))
     verts = Vector{T}()
+    S = Vector{T}()
     for v in vertices(g)
         vcolor[v] != 0 && continue
-        S = Vector{T}([v])
+        push!(S, v)
         vcolor[v] = 1
         while !isempty(S)
             u = S[end]
-            w = 0
-            for n in outneighbors(g, u)
-                if vcolor[n] == 1
-                    error("The input graph contains at least one loop.") # TODO 0.7 should we use a different error?
-                elseif vcolor[n] == 0
-                    w = n
-                    break
+            if vcolor[u] == 1
+                vcolor[u] = 2
+                for n in outneighbors(g, u)
+                    if vcolor[n] == 2
+                        error("The input graph contains at least one loop.") # TODO 0.7 should we use a different error?
+                    elseif vcolor[n] == 0
+                        vcolor[n] = 1
+                        push!(S, n)
+                    end
                 end
-            end
-            if w != 0
-                vcolor[w] = 1
-                push!(S, w)
             else
-                vcolor[u] = 2
-                push!(verts, u)
+                vcolor[u] = 3
                 pop!(S)
+                pushfirst!(verts, u)
```

Comment:
How are we sure that each vertex lands in `verts` exactly once?

Reply:
Each vertex is visited at least once (outer loop). Each vertex is visited at most once (added to vertex_heap only if color is 0).
Then vertex is visited once to explore its unvisited subtree, and color goes to 2. Next time it is visited, its color goes to 3, and vertex is removed from queue and added to verts, so not visited anymore.